mirror_maker.go

```diff
@@ -209,13 +264,74 @@ func (this *MirrorMaker) produceRoutine(producer Producer, channelIndex int) {
 	partitionEncoder := &Int32Encoder{}
 	for msg := range this.messageChannels[channelIndex] {
 		if this.config.PreservePartitions {
-			producer.Input() <- &ProducerMessage{Topic: this.config.TopicPrefix + msg.Topic, Key: uint32(msg.Partition), Value: msg.Value, KeyEncoder: partitionEncoder}
+			producer.Input() <- &ProducerMessage{Topic: this.config.TopicPrefix + msg.Topic, Key: uint32(msg.Partition), Value: msg.DecodedValue, KeyEncoder: partitionEncoder}
+		} else {
+			producer.Input() <- &ProducerMessage{Topic: this.config.TopicPrefix + msg.Topic, Key: msg.Key, Value: msg.DecodedValue}
+		}
+	}
+}
+
+func (this *MirrorMaker) timingsRoutine(producer Producer) {
+	var keyDecoder Decoder
+	if this.config.PreservePartitions {
+		keyDecoder = &Int32Decoder{}
+	} else {
+		keyDecoder = this.config.KeyDecoder
+	}
+
+	for msg := range producer.Successes() {
+		decodedKey, err := keyDecoder.Decode(msg.Key.([]byte))
+		if err != nil {
+			Errorf(this, "Failed to decode %v", msg.Key)
+		}
+		decodedValue, err := this.config.ValueDecoder.Decode(msg.Value.([]byte))
+		if err != nil {
+			Errorf(this, "Failed to decode %v", msg.Value)
+		}
+
+		if record, ok := decodedValue.(*avro.GenericRecord); ok {
+			record = this.addTiming(record)
+			if this.config.PreservePartitions {
+				this.timingsProducer.Input() <- &ProducerMessage{Topic: "timings_" + msg.Topic, Key: int32(decodedKey.(uint32)), Value: record}
+			} else {
+				this.timingsProducer.Input() <- &ProducerMessage{Topic: "timings_" + msg.Topic, Key: decodedKey, Value: record}
+			}
 		} else {
-			producer.Input() <- &ProducerMessage{Topic: this.config.TopicPrefix + msg.Topic, Key: msg.Key, Value: msg.Value}
+			Errorf(this, "Invalid avro schema type %s", decodedValue)
 		}
 	}
 }
 
+func (this *MirrorMaker) failedRoutine(producer Producer) {
+	for msg := range producer.Errors() {
+		Error("mirrormaker", msg.err)
+	}
+}
+
+func (this *MirrorMaker) addTiming(record *avro.GenericRecord) *avro.GenericRecord {
+	now := time.Now().Unix()
```

Comment:
This is only measuring when the timestamp is generated which may not be the same timestamp of the event that the calling application wanted you to track and add timing to the avro for it. e.g. ad server impression comes in to request, then to response. There would be 4 timings in the structure (at least). 1) the time stamp of the ad request impression 2) right before the serialization for that last event happened 3) the timestamp of when the response happened 4) a timestamp for when the actual serialization event occured. If you want to create a data structure for this you can/should so we know that the "atomic time" was the time for the event which sometimes is the ser/de event but not always.